firmware/application/apps/ui_text_editor.cpp

```diff
@@ -221,28 +222,45 @@ void TextViewer::paint_text(Painter& painter, uint32_t line, uint16_t col) {
                  r.top() + (int)i * char_height,
                  clear_width * char_width, char_height},
                 style().background);
+
+        // if cursor line got overwritten, disable XOR of old cursor when displaying new cursor
+        if (i == paint_state_.line)
+            paint_state_.line = UINT32_MAX;
     }
 }
 
 void TextViewer::paint_cursor(Painter& painter) {
     if (!has_focus())
         return;
 
-    auto draw_cursor = [this, &painter](uint32_t line, uint16_t col, Color c) {
+    auto xor_cursor = [this, &painter](int32_t line, uint16_t col) {
         auto r = screen_rect();
-        line = line - paint_state_.first_line;
-        col = col - paint_state_.first_col;
-
-        painter.draw_rectangle(
-            {(int)col * char_width - 1,
-             r.top() + (int)line * char_height,
-             char_width + 1, char_height},
-            c);
+        int cursor_width = char_width + 1;
+
+        int x = (col - paint_state_.first_col) * char_width - 1;
+        if (x < 0) {  // cursor is one pixel narrower when in left column
+            cursor_width--;
+            x = 0;
+        }
+
+        int y = r.top() + (line - paint_state_.first_line) * char_height;
+
+        int cursor_pixels = cursor_width * char_height;
+        ColorRGB888 pbuf888[cursor_pixels];
```

Comment:
Does that work? Doesn't that have to be dynamically allocated?

Reply:
You can use a std::vector instead, but man, it would be so much better to not have to dyn-alloc here. Why is there the weirdness with the first column? Maybe we could fix the special case and then you could use a static buffer here.

Reply:
Maybe shift the text over 1 px on line 214? Or does that just give you the same problem on the other side when you have an evenly divisible width?

Reply:
Oh neat, VLAs are a g++ extension that use alloca under the hood. Cool 👍

Reply:
I didn't care for this ~1KB allocation either, and was thinking of doing one row at a time instead and I suppose a static buffer.  (Actually I could do one pixel at a time and avoid allocation entirely, but I think it would be much much slower calling read_pixels & draw_pixels for one pixel at a time.  Calling it 9 or 17 times might not be too bad.)